File: CODE/OOviewOfBoxscoreData.py
# Copyright 2025 by Steven J. Keyes. All rights reserved.
# FILE: oo_view_of_boxscore_data.py
# DATE 2025-11-13
# DESCRIPTION: Python equivalent of OOviewOfBoxscoreData.py

import sqlite3
import os
from dataclasses import dataclass, field
from typing import List
import pandas as pd

@dataclass
class Player:
    """Player class representing individual player statistics"""
    player_id: int
    fgm: int
    fga: int
    fg3m: int
    fg3a: int
    ftm: int
    fta: int


@dataclass
class Team:
    """Team class containing a list of players"""
    team_id: int
    players: List[Player] = field(default_factory=list)


@dataclass
class Game:
    """Game class containing teams that played"""
    game_id: int
    teams: List[Team] = field(default_factory=list)


@dataclass
class Season2025:
    """Season class containing all games"""
    season_id: int
    games: List[Game] = field(default_factory=list)


def load_season_from_db(db_path: str) -> dict:
    """
    Load data from SQLite database and create Season2025 object
    
    Args:
        db_path: Path to the SQLite database file
        
    Returns:
        Dictionary containing season object and reference tables
    """
    # Connect to the database
    con = sqlite3.connect(db_path)
    
    # Load all tables
    boxscores = pd.read_sql_query("SELECT * FROM Boxscores", con)
    teams_table = pd.read_sql_query("SELECT * FROM Teams", con)
    players_table = pd.read_sql_query("SELECT * FROM Players", con)
    season_table = pd.read_sql_query("SELECT * FROM Season2025", con)
    
    # Close the connection
    con.close()
    
    # Get unique game IDs
    game_ids = boxscores['GAME_ID'].unique()
    
    # Create list to store games
    games_list = []
    
    # Process each game
    for game_id in game_ids:
        # Get all boxscore entries for this game
        game_boxscores = boxscores[boxscores['GAME_ID'] == game_id]
        
        # Get unique team IDs in this game
        team_ids = game_boxscores['TEAM_ID'].unique()
        
        # Create list to store teams
        teams_list = []
        
        # Process each team
        for team_id in team_ids:
            # Get all players for this team in this game
            team_boxscores = (
                game_boxscores[game_boxscores['TEAM_ID'] == team_id])
            
            # Create list to store players
            players_list = []
            
            # Process each player
            for _, row in team_boxscores.iterrows():
                player = Player(
                    player_id=int(row['PLAYER_ID']),
                    fgm=int(row['FGM']),
                    fga=int(row['FGA']),
                    fg3m=int(row['FG3M']),
                    fg3a=int(row['FG3A']),
                    ftm=int(row['FTM']),
                    fta=int(row['FTA'])
                )
                players_list.append(player)
            
            # Create Team object
            team = Team(
                team_id=int(team_id),
                players=players_list
            )
            teams_list.append(team)
        
        # Create Game object
        game = Game(
            game_id=int(game_id),
            teams=teams_list
        )
        games_list.append(game)
    
    # Create Season2025 object
    season = Season2025(
        season_id=2025,
        games=games_list
    )
    
    # Return season object and the additional tables for reference
    return {
        'season': season,
        'teams_info': teams_table,
        'players_info': players_table,
        'season_stats': season_table
    }


def get_player_name(player_id: int, players_info: pd.DataFrame) -> str:
    """Get player name from players_info DataFrame"""
    player_mask = players_info['PLAYER_ID'] == player_id
    result = players_info[player_mask]['PLAYER_NAME']
    return result.iloc[0] if len(result) > 0 else None


def get_team_name(team_id: int, teams_info: pd.DataFrame) -> str:
    """Get team name from teams_info DataFrame"""
    result = (
        teams_info[teams_info['TEAM_ID'] == team_id]['TEAM_NAME'])
    return result.iloc[0] if len(result) > 0 else None


def main():
    """Main function to demonstrate the data loading and access"""
    # Get database path
    examples_dir = os.getenv("EXAMPLES", "")
    path_to_database = os.path.join(examples_dir, "Boxscores.db")
    
    # Load data
    data = load_season_from_db(path_to_database)
    
    season = data['season']
    teams_info = data['teams_info']
    players_info = data['players_info']
    season_stats = data['season_stats']
    
    # Access the first game
    first_game = season.games[0]
    print('--- First game ---')
    print(first_game)
    print()
    
    # Access the first team in the first game
    first_team = first_game.teams[0]
    print('--- First team ---')
    print(first_team)
    print()
    
    # Access players in that team
    first_player = first_team.players[0]
    print('--- First player ---')
    print(first_player)
    print()
    
    # Get player name
    player_name = get_player_name(first_player.player_id, players_info)
    print('--- Player name ---')
    print(player_name)
    print()
    
    # Get team name
    team_name = get_team_name(first_team.team_id, teams_info)
    print('--- Team name ---')
    print(team_name)
    print()


if __name__ == "__main__":
    main()

File: CODE/OOviewOfBoxscoreData.R
# Copyright 2025 by Steven J. Keyes. All rights reserved.
# FILE: OOviewOfBoxscoreData.R
# DATE 2025-11-13
# DESCRIPTION: 
# This R program is a data modeling and transformation tool. 
# Its main objective is to read flat data tables from a database 
# and reconstruct them into a structured, hierarchical collection 
# of R objects that mimic the physical world of a basketball 
# season (Season -> Game -> Team -> Player).

library(DBI)
library(RSQLite)
library(S7)

# Player Class
Player <- new_class(
  name = "Player",
  properties = list(
    player_id = class_integer,
    fgm = class_integer,
    fga = class_integer,
    fg3m = class_integer,
    fg3a = class_integer,
    ftm = class_integer,
    fta = class_integer
  )
)

# Team Class
Team <- new_class(
  name = "Team",
  properties = list(
    team_id = class_integer,
    players = class_list
  )
)

# Game Class
Game <- new_class(
  name = "Game",
  properties = list(
    game_id = class_integer,
    teams = class_list
  )
)

# Season2025 Class
Season2025 <- new_class(
  name = "Season2025",
  properties = list(
    season_id = class_integer,
    games = class_list
  )
)

# Function to load data from SQLite database and create Season2025 object
load_season_from_db <- function(db_path) {
  # Connect to the database
  con <- dbConnect(SQLite(), db_path)
  
  # Load all tables
  boxscores <- dbReadTable(con, "Boxscores")
  teams_table <- dbReadTable(con, "Teams")
  players_table <- dbReadTable(con, "Players")
  season_table <- dbReadTable(con, "Season2025")
  
  # Close the connection
  dbDisconnect(con)
  
  # Get unique game IDs
  game_ids <- unique(boxscores$GAME_ID)
  
  # Create list to store games
  games_list <- list()
  
  # Process each game
  for (game_id in game_ids) {
    
    # Get all boxscore entries for this game
    game_boxscores <- boxscores[boxscores$GAME_ID == game_id, ]
    
    # Get unique team IDs in this game
    team_ids <- unique(game_boxscores$TEAM_ID)
    
    # Create list to store teams
    teams_list <- list()
    
    # Process each team
    for (team_id in team_ids) {
      # Get all players for this team in this game
      team_boxscores <- game_boxscores[game_boxscores$TEAM_ID == team_id, ]
      
      # Create list to store players
      players_list <- list()
      
      # Process each player
      for (i in 1:nrow(team_boxscores)) {
        player <- Player(
          player_id = as.integer(team_boxscores$PLAYER_ID[i]),
          fgm = as.integer(team_boxscores$FGM[i]),
          fga = as.integer(team_boxscores$FGA[i]),
          fg3m = as.integer(team_boxscores$FG3M[i]),
          fg3a = as.integer(team_boxscores$FG3A[i]),
          ftm = as.integer(team_boxscores$FTM[i]),
          fta = as.integer(team_boxscores$FTA[i])
        )
        players_list[[i]] <- player
      }
      
      # Create Team object
      team <- Team(
        team_id = as.integer(team_id),
        players = players_list
      )
      teams_list[[length(teams_list) + 1]] <- team
    }
    
    # Create Game object
    game <- Game(
      game_id = as.integer(game_id),
      teams = teams_list
    )
    games_list[[length(games_list) + 1]] <- game
  }
  
  # Create Season2025 object
  season <- Season2025(
    season_id = 2025L,
    games = games_list
  )
  
  # Return season object and the additional tables for reference
  return(list(
    season = season,
    teams_info = teams_table,
    players_info = players_table,
    season_stats = season_table
  ))
}

# Function to get player name from players_info
get_player_name <- function(player_id, players_info) {
  players_info$PLAYER_NAME[players_info$PLAYER_ID == player_id]
}

# Function to get team name from teams_info
get_team_name <- function(team_id, teams_info) {
  teams_info$TEAM_NAME[teams_info$TEAM_ID == team_id]
}

path_to_database <- file.path(Sys.getenv("EXAMPLES"), "Boxscores.db")
data <- load_season_from_db(path_to_database)

season <- data$season
teams_info <- data$teams_info
players_info <- data$players_info
season_stats <- data$season_stats

# Access the first game
first_game <- season@games[[1]]
print('--- First game ---')
print(first_game)
cat('\n')

# Access the first team in the first game
first_team <- first_game@teams[[1]]
print('--- First team ---')
print(first_team)
cat('\n')

# Access players in that team
first_player <- first_team@players[[1]]
print('--- First player ---')
print(first_player)
cat('\n')

# Get player name
player_name <- get_player_name(first_player@player_id, players_info)
print('--- Player name ---')
print(player_name)
cat('\n')

# Get team name
team_name <- get_team_name(first_team@team_id, teams_info)
print('--- Team name ---')
print(team_name)
cat('\n')
